Remove and lint unused Redux action types

aduth opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

## Proposed Changes

* Removes unused Redux action types
* Adds lint script to ensure that all defined action types are in use

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Action types are included in the main application bundle and their values cannot be minified, so unused action types unnecessarily inflate bundle size

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Confirm there is no usage of removed action types
* Verify that the included lint script produces a non-zero exit code if reintroducing unused action types (e.g. after `git checkout trunk -- client/state/action-types.ts`)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aduth committed:
Remove unused Redux action types

aduth committed:
Add linter to check for unused action types

aduth commented on bin/find-unused-state-action-types.sh:
```diff
@@ -0,0 +1,16 @@
+    'grep -rq --include="*/actions/*" \
+                --include="reducer.js" \
+                --include="reducer.ts" \
+                --include="actions.js" \
+                --include="actions.ts" \
```
I feel like we shouldn't need to include these, as action types that are only used in action dispatching (i.e. not referenced by reducers or middlewares) aren't actually doing anything to cause state changes? Or am I mistaken?

There's a lot of extra action types that are flagged when not including these patterns, though my initial impression is that they (and the actions that dispatch them) could be safely removed. I'd want some extra eyes on that, and it's a big enough effort to split out to a separate pull request.

Flagged action types after removing the actions-related `--include` flags:

**Edit:** There's a few more exceptions in this list like those referenced through `data-layer` or variations on `reducer.ts` (`reducers.ts`), but the main point stands.

```
Unexpected unused action type: ACCOUNT_RECOVERY_SETTINGS_RESEND_VALIDATION_SUCCESS
Unexpected unused action type: ACCOUNT_RECOVERY_SETTINGS_RESEND_VALIDATION_FAILED
Unexpected unused action type: ANALYTICS_MULTI_TRACK
Unexpected unused action type: COMMAND_PALETTE_OPEN
Unexpected unused action type: COMMAND_PALETTE_CLOSE
Unexpected unused action type: CURRENT_USER_FETCH
Unexpected unused action type: EXPERIMENT_ASSIGNMENT_RECEIVE
Unexpected unused action type: EXPORT_STATUS_FETCH
Unexpected unused action type: GRAVATAR_RECEIVE_IMAGE_FAILURE
Unexpected unused action type: HOSTING_SFTP_USER_CREATE
Unexpected unused action type: HOSTING_PHP_VERSION_REQUEST
Unexpected unused action type: HOSTING_PHP_VERSION_SET_REQUEST
Unexpected unused action type: HOSTING_SFTP_PASSWORD_RESET
Unexpected unused action type: HOSTING_RESTORE_DATABASE_PASSWORD
Unexpected unused action type: HOSTING_SSH_ACCESS_DISABLE
Unexpected unused action type: HOSTING_SSH_ACCESS_ENABLE
Unexpected unused action type: HOSTING_WP_VERSION_SET_REQUEST
Unexpected unused action type: INVITE_ACCEPTED
Unexpected unused action type: JETPACK_PARTNER_PORTAL_LICENSE_COUNTS_REQUEST
Unexpected unused action type: JETPACK_SCAN_THREAT_FIX
Unexpected unused action type: JETPACK_SCAN_THREAT_UNIGNORE
Unexpected unused action type: JETPACK_SCAN_THREAT_IGNORE
Unexpected unused action type: JETPACK_SCAN_THREATS_FIX_ALL
Unexpected unused action type: JETPACK_SCAN_THREATS_GET_FIX_STATUS
Unexpected unused action type: JETPACK_SITE_ALERT_THREAT_UNIGNORE
Unexpected unused action type: JETPACK_SITE_ALERT_THREAT_FIX
Unexpected unused action type: JETPACK_SITE_ALERT_THREAT_IGNORE
Unexpected unused action type: JITM_OPEN_HELP_CENTER
Unexpected unused action type: KEYRING_CONNECTION_DELETE_FAILURE
Unexpected unused action type: LASAGNA_SOCKET_DISCONNECTED
Unexpected unused action type: MAILCHIMP_SETTINGS_UPDATE_FAILURE
Unexpected unused action type: MEMBERSHIPS_COUPON_ADD
Unexpected unused action type: MEMBERSHIPS_COUPON_ADD_FAILURE
Unexpected unused action type: MEMBERSHIPS_COUPON_UPDATE
Unexpected unused action type: MEMBERSHIPS_COUPON_DELETE_FAILURE
Unexpected unused action type: MEMBERSHIPS_COUPON_UPDATE_FAILURE
Unexpected unused action type: MEMBERSHIPS_GIFT_DELETE_FAILURE
Unexpected unused action type: MEMBERSHIPS_GIFT_DELETE
Unexpected unused action type: MEMBERSHIPS_GIFT_ADD_FAILURE
Unexpected unused action type: MEMBERSHIPS_GIFT_ADD
Unexpected unused action type: MEMBERSHIPS_PRODUCT_ADD_FAILURE
Unexpected unused action type: MEMBERSHIPS_PRODUCT_ADD
Unexpected unused action type: MEMBERSHIPS_PRODUCT_UPDATE
Unexpected unused action type: MEMBERSHIPS_PRODUCT_DELETE_FAILURE
Unexpected unused action type: MEMBERSHIPS_PRODUCT_UPDATE_FAILURE
Unexpected unused action type: NOTIFICATION_SETTINGS_UPDATE
Unexpected unused action type: POST_RESTORE_SUCCESS
Unexpected unused action type: POST_SAVE_FAILURE
Unexpected unused action type: PUBLICIZE_CONNECTION_CREATE_FAILURE
Unexpected unused action type: PUBLICIZE_CONNECTION_DELETE_FAILURE
Unexpected unused action type: PUBLICIZE_CONNECTION_UPDATE_FAILURE
Unexpected unused action type: PUBLICIZE_CONNECTIONS_REQUEST_SUCCESS
Unexpected unused action type: PUSH_NOTIFICATIONS_API_NOT_READY
Unexpected unused action type: SITE_KEYRINGS_DELETE_FAILURE
Unexpected unused action type: SITE_KEYRINGS_UPDATE_FAILURE
Unexpected unused action type: SITE_LAUNCH_SUCCESS
Unexpected unused action type: SITE_LAUNCH_FAILURE
Unexpected unused action type: EMAIL_STATS_REQUEST_SUCCESS
```

tyxla replied:
Yeah, data layer is one example: you could trigger an action with one of these action types, and it would trigger a request, but not necessarily have a corresponding state tree. I personally don't mind having them all declared here. Whether we want to keep the data layer, that's a topic for another conversation, but some of us think removing it may be feasible.

If it's not related to the data layer, and we are dispatching an action that doesn't have a corresponding reducer...it might be a signal that these actions shouldn't be dispatched altogether and can be removed.

aduth commented:
> This PR does not affect the size of JS and CSS bundles shipped to the user's browser.

Hmm, guessing maybe these are actually effectively tree-shaken after all?

I think it could still be valuable to keep these consistent in terms of code size, bundler work effort, and developer understanding of the scope of action types available.

tyxla commented on package.json:
```diff
@@ -103,6 +103,7 @@
 		"lint:css": "stylelint \"**/*.{css,scss}\"",
 		"lint:js": "eslint --ext .js,.jsx,.ts,.tsx,.mjs,.json --cache .",
 		"lint:mixedindent": "mixedindentlint --ignore-comments \"client/**/*.scss\" \"assets/**/*.scss\" \"!build/**\" \"!node_modules/**\" \"!public/**\"",
+		"lint:unused-state-action-types": "node bin/find-unused-state-action-types.sh",
```
This way, this script will only run in CI, right? Any easy way to make it run locally and display linting errors in the IDE?

aduth replied:
> This way, this script will only run in CI, right? Any easy way to make it run locally and display linting errors in the IDE?

Yeah, as currently implemented, I'm expecting this will most often be flagged through CI, though technically you can run the lint script locally as well.

An editor integration would be nice to have. I'd imagine that'd be implemented as a custom ESLint rule? The proposed direction felt like a good middle-ground of enforcing an expectation without going too deep into a rabbit hole about how to surface it. It's likely to be a pretty uncommon scenario, and addressing a CI failure wouldn't be overly burdensome IMO. I'm also conscious of performance here since even the low-level `grep`'ing was surprisingly slow without parallelization.

tyxla commented on bin/find-unused-state-action-types.sh:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+grep -r "export const [A-Z_]* =" client/state/action-types.ts | \
```
Let's add a comment that explains what we're aiming to do.

aduth replied:
> Let's add a comment that explains what we're aiming to do.

Added comment in 827e8f2fbbb.

tyxla commented on bin/find-unused-state-action-types.sh:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+grep -r "export const [A-Z_]* =" client/state/action-types.ts | \
```
Should we have a check to ensure the action types file exists before doing anything else? Could be helpful if someone removes or moves the file without knowing about this script.

aduth replied:
> Should we have a check to ensure the action types file exists before doing anything else? Could be helpful if someone removes or moves the file without knowing about this script.

Updated in e95a8a265b6 to fail script if file is missing.

tyxla commented on bin/find-unused-state-action-types.sh:
```diff
@@ -0,0 +1,16 @@
+grep -r "export const [A-Z_]* =" client/state/action-types.ts | \
+    sed -E 's/.*export const ([A-Z_]+) =.*/\1/' | \
+    xargs -n 1 -P 8 -I{} sh -c \
+    'grep -rq --include="*/actions/*" \
```
It may be reasonable to declare the list of patterns in a variable at the top.

aduth replied:
> It may be reasonable to declare the list of patterns in a variable at the top.

I was able to clean this up a little bit in 39f83571138. Per the recommendation in https://github.com/Automattic/wp-calypso/pull/102908#issuecomment-2851432217, I'm also expecting we'll be able to drop the two `actions`-related `--include` flags in the next pull request.

jsnajdr commented:
The linting could be implemented as a webpack plugin. Webpack has a precise and authoritative information on usage of each export. A webpack plugin could walk through the `action-types` module and flag the exports that are not used.

The `action-types` module is processed by the `InlineConstantExportsPlugin` which inlines the string at places of use without using the import/export machinery. It could be an additional feature of this plugin to find the unused exports.

aduth commented:
> The linting could be implemented as a webpack plugin. Webpack has a precise and authoritative information on usage of each export. A webpack plugin could walk through the `action-types` module and flag the exports that are not used.
> 
> The `action-types` module is processed by the `InlineConstantExportsPlugin` which inlines the string at places of use without using the import/export machinery. It could be an additional feature of this plugin to find the unused exports.

That's an interesting idea. I'm having some trouble getting the Webpack plugin to work as I'd expect it to in my local environment, and even in production I'm not seeing the intended effect of inlined constants 🤔  Is there some trick to get it to do the inlining?

I'm also a little hesitant to mix build tooling and testing like this, as in... what would be the expected outcome if the Webpack plugin discovers an unused constant. Should it raise an error and fail the build? Do we want to risk build interruption like that? If it just logs a warning, that'll surely be overlooked or ignored.

Happy to continue investing in this, though unsure how much value we'll get from it. I was hoping the added checks here would be a pretty low-touch improvement, with the alternative being to just fix the offending lines in `action-types.ts` and move on without any sort of regression checking. 

<details><summary>Work-in-progress patch</summary>
<pre><code>diff --git a/client/webpack.config.js b/client/webpack.config.js
index 97ab90e5672..449a59ddf55 100644
--- a/client/webpack.config.js
+++ b/client/webpack.config.js
@@ -357 +357,5 @@ const webpackConfig = {
-		new InlineConstantExportsPlugin( /\/client\/state\/action-types.js$/ ),
+		new InlineConstantExportsPlugin( /\/client\/state\/action-types.js$/, {
+			onUnusedConstant: ( constant ) => {
+				throw new Error( `Unexpected unused constant: ${ constant }` );
+			},
+		} ),
diff --git a/packages/webpack-inline-constant-exports-plugin/index.js b/packages/webpack-inline-constant-exports-plugin/index.js
index b39aceddef3..12800d57b6a 100644
--- a/packages/webpack-inline-constant-exports-plugin/index.js
+++ b/packages/webpack-inline-constant-exports-plugin/index.js
@@ -11,0 +12,5 @@ function addConstantExport( module, name, value ) {
+function markConstantUsed( module, name ) {
+	module.usedConstantExports ??= new Set();
+	module.usedConstantExports.add( name );
+}
+
@@ -17,0 +23,2 @@ class InlineConstantExportsPlugin {
+	constantsModules = new Set();
+
@@ -22 +29 @@ class InlineConstantExportsPlugin {
-	constructor( matchers ) {
+	constructor( matchers, options ) {
@@ -27,0 +35,2 @@ class InlineConstantExportsPlugin {
+
+		this.options = options ?? {};
@@ -49,0 +59,2 @@ class InlineConstantExportsPlugin {
+						this.constantsModules.add( module );
+
@@ -199,0 +211,6 @@ class InlineConstantExportsPlugin {
+
+									// Mark the constant as used, if the plugin consumer is
+									// interested in unused constants.
+									if ( this.options.onUnusedConstant ) {
+										markConstantUsed( module, depId );
+									}
@@ -222,0 +240,10 @@ class InlineConstantExportsPlugin {
+
+					if ( this.options.onUnusedConstant ) {
+						this.constantsModules.forEach( ( module ) => {
+							for ( const depId in module.constantExports ) {
+								if ( ! module.usedConstantExports.has( depId ) ) {
+									this.options.onUnusedConstant( depId );
+								}
+							}
+						} );
+					}
</code></pre></details>

tyxla commented:
My $0.02: Webpack can do and has been doing a relatively good job of keeping unused code out of our bundles, but ideally, I'd prefer not to keep unused code around at all.

jsnajdr commented:
> I'm having some trouble getting the Webpack plugin to work as I'd expect it to in my local environment, and even in production I'm not seeing the intended effect of inlined constants

I'm now testing it too, and it seems that the plugin got broken silently at some moment in the last 6 years 🙂

jsnajdr commented:
About whether to use webpack or not: I just intuitively don't like the new script very much, because:
- it adds a new linting task for something that's not very important, for me it doesn't pass the threshold for adding a new check. In the future someone will need to maintain it, try to understand what's going on there etc.
- it feels fragile: hardcoded paths and file names, using regexp to analyze syntax.

As opposed to that, webpack (and maybe also the TypeScript complilation) already has a full and precise info about all the dependencies and their usage.

The only downside of the webpack check is that we in fact have multiple webpack compilations (Oddyssey Stats etc.) When one of them doesn't use a certain Redux action, some other compilation can still use it. It's not OK if the first compilation reports it as unused.

jsnajdr commented:
> seems that the plugin got broken silently at some moment 

Of course, the file was renamed from `action-types.js` to `action-types.ts` in 2021 (#51717) and since then the plugin can't match it 🤦

tyxla commented:
> > seems that the plugin got broken silently at some moment
> 
> Of course, the file was renamed from `action-types.js` to `action-types.ts` in 2021 (#51717) and since then the plugin can't match it 🤦

We need to get better at spotting that kind of regressions 🤦

aduth commented:
I don't disagree that this is not a critical thing to be checking. Part of the reason for this particular approach was that it didn't feel worthwhile to spend a large chunk of time trying to integrate a proper check, and this was pretty low-effort. I didn't want to just fix the issues and move on, since this problem seems highly likely to reoccur in the future, particularly as Redux features may be removed or migrated to React Redux implementations. I also think there's naturally going to be some maintenance requirement for any of our tests, as evidenced by the same sort of fragility of hard-coded path names that you mentioned being a downside of this approach being the same that broke the Webpack plugin.

I'm not particularly resistant to revising the approach here or dropping it altogether, but I think at a high level if we set too high of expectations and level of effort for this kind of regression testing, it's likely that most people would choose to not do it at all.

Maybe that's okay in this case, though. 🤷 Particularly since it didn't have the optimization benefit I was hoping for.

tyxla commented:
I guess In the meantime, we can just delete the unused action type exports and move on after fixing the file name issue, letting webpack do its own thing and shake some trees.

aduth committed:
Include comment explaining purpose of script

See: https://github.com/Automattic/wp-calypso/pull/102908/files#r2068249168

aduth committed:
Simplify grep to avoid including file name in result

aduth committed:
Fail command if file not found

See: https://github.com/Automattic/wp-calypso/pull/102908/files#r2068250538

aduth committed:
Consolidate extension matching for JS, TS

Help readability

Related: https://github.com/Automattic/wp-calypso/pull/102908/files#r2068252242

aduth committed:
Exempt data-layer references to action types

aduth commented:
> What I would propose is to keep the script (with feedback revisions) and the action types removal, and then in a follow-up pull request revise the script to remove the exceptions for `actions.js`/`actions.ts` to identify more substantial code changes.

See #103138 for preliminary follow-up work.
